### ISSUE TITLE
Problem of results on MetaShift

### ISSUE THREAD

**Issue by chandayh** [Wuyxin/DISC#1]
Dear authors,
Thank you so much for your work! I have tried the codes on your [scripts](https://github.com/Wuyxin/DISC#run-the-training-process) on both the ERM and DISC models. Here is what I exactly ran: 
```
SEED=0
ROOT=/data/data/DISC # Set your code root here
python run_expt.py \
-shift_type confounder -dataset MetaDatasetCatDog -target_name cat -confounder_names background --lr 0.001 --batch_size 16 \
--weight_decay 0.0001 --model resnet50 --n_epochs 100 --log_dir $ROOT/output/ \
--root_dir $ROOT/data/metashifts/MetaDatasetCatDog --save_best --save_last --seed $SEED 

SEED=0
N_CLUSTERS=2
ROOT=/data/data/DISC # Set your code root here
CUDA_VISIBLE_DEVICES=1 python run_expt.py \
--shift_type confounder --dataset MetaDatasetCatDog --target_name cat --confounder_names background \
--lr 0.0005 --batch_size 16 \
--weight_decay 0.0001 --model resnet50 --n_epochs 100  --log_dir $ROOT/output/ \
--root_dir $ROOT/data/metashifts/MetaDatasetCatDog \
--erm_path output/MetaDatasetCatDog/ERM/reweight_groups=0-augment=0-lr=0.001-batch_size=16-n_epochs=100-seed=0/best_model.pth \
--concept_img_folder $ROOT/synthetic_concepts --concept_categories everything \
--n_clusters $N_CLUSTERS --augment_data --save_last --save_best --seed $SEED --disc 
```
However from the results that are shown in the output, (as shown below), it is quite different from in the table 1, where the ERM should achieve 72.9% and DISC should achieve 75.5% in avg acc. 
```
ERM:
worst_group_acc     0.686567
mean_differences    0.155132
group_avg_acc       0.764133
avg_acc             0.773913

DISC:
worst_group_acc     0.687259
mean_differences    0.034134
group_avg_acc       0.704326
avg_acc             0.702174
```
Could you please help me with it? Also do you have the pretrained weights on ERM and DISC and corresponding eval results?

Thanks,

**Comment by Wuyxin** [comment]
Hi,

Thanks for the interests!

I pushed the example outputs and model weights on MetaShifts (model weights are kinda big. I will remove them from repo later and find other ways to upload).

My observation is that the results on MetaShifts are very stable. But that doesn't exclude scenarios like bad clustering or bad seeds. I would suggest you try to run more random seeds, also compare the loss in the example outputs to see if anything is significantly different.

Let me know hoe it goes and I am happy to help more!

**Comment by chandayh** [comment]
Thank you for your effort! However, I have used both of the model weights uploaded with respect to ERM and DISC, and below are the results:
```
ERM:
worst_group_acc     0.644788
mean_differences    0.245760
group_avg_acc       0.767667
avg_acc             0.752174

DISC:
worst_group_acc     0.749035
mean_differences    0.027085
group_avg_acc       0.762577
avg_acc             0.760870
```
The worst group accs are similar to what is shown in table 1, but there's still discrepancy in average acc. I am wondering, could it be the problem of the size of the generated dataset? Basically I saw that the generated dataset `(os.listdir('synthetic_concepts')` has the length 224, while the number of items in `synthetic_concepts/metadata.json` is 235, which means there are duplicate items. Could this be the problem? Also I will try experiments with other random seeds.

**Comment by Wuyxin** [comment]
Sorry, I didn't understand. Which discrepancy are you referring to? I think the avg results here are higher than the reported average performance in our paper.

There are some shared concepts among different concept categories, but these concepts won't be duplicated during training, since we always take the unique set of concepts before training.

**Comment by chandayh** [comment]
Thank you for explanation on the generated synthetic concepts. 
By mentioning discrepancy, I mean the avg acc of ERM is 0.752174 and the avg acc of DISC is 0.760870. This is quite different from in the paper (72.9 to 75.5), since DISC doesn't help 'so' much on the pretrained ERM. I am wondering why this is the case.

**Comment by Wuyxin** [comment]
The version I pushed is ERM+aug instead of ERM. 

And the performance of ERM+aug is 75.5 ± 1.7% on avg acc.

**Comment by chandayh** [comment]
Alright, that makes sense. Thank you!

**Comment by Wuyxin** [comment]
No worries! Let me know if you have further question about reproducing the results :)

**Comment by chandayh** [comment]
Thank you, I will conduct experiments using different random seeds.

**Comment by chandayh** [comment]
Hello, I have tried several random seeds and used ERM as pretrained, DISC as fine tuning. Below is a table for results:
<html>
<body>
<b style="font-weight:normal;" id="docs-internal-guid-3c891e42-7fff-92e8-418e-e0aab178c5a2"><div dir="ltr" style="margin-left:0pt;" align="left">

ERM | worst_group_acc | mean_differences | group_avg_acc | avg_acc
-- | -- | -- | -- | --
seed0 | 0.686567 | 0.155132 | 0.764133 | 0.773913
seed1 | 0.583012 | 0.292610 | 0.729317 | 0.710870
seed2 | 0.552124 | 0.368275 | 0.736261 | 0.713044
seed3 | 0.667954 | 0.187768 | 0.761838 | 0.750000

</div></b>
</body>
</html>

<html>
<body>
<b style="font-weight:normal;" id="docs-internal-guid-52b304af-7fff-caa4-df7a-44edd6ba35dc"><div dir="ltr" style="margin-left:0pt;" align="left">

DISC | worst_group_acc | mean_differences | group_avg_acc | avg_acc
-- | -- | -- | -- | --
seed0 | 0.687259 | 0.034134 | 0.704326 | 0.702174
seed1 | 0.640927 | 0.145143 | 0.713498 | 0.704348
seed2 | 0.733591 | 0.057454 | 0.762318 | 0.758696
seed3 | 0.679537 | 0.066732 | 0.712903 | 0.708696

</div></b>
</body>
</html>

In the table I found out that the worst group acc is always better using DISC, but the avg acc is not always this case. Could you please share some hints on achieving the results in the article? Specifically my results of DISC is not so good as reported in the article.

**Comment by Wuyxin** [comment]
Hmm the results are quite different from what I got. Although I won't be surprised that ERM may have better avg ACC in some cases (lots of domain adaptation methods underperform ERM on avg metric https://wilds.stanford.edu/leaderboard/)

It seems to me that the problem goes beyond reproducing DISC since the ERM's performance is also not sufficient. I pushed the ERM model to the repo as well as the `filtered` folder under DISC output if that helps. Just to check, did you use the same `filtered` folder across these DISC runs or you regenerate it everytime? It may make a difference as well.

**Comment by chandayh** [comment]
Yeah I used the same `filtered` folder since [this line](https://github.com/Wuyxin/DISC/blob/25717ffe3ae7eb74056a23936ccbfa4a84815535/disc/concept_utils/filter_concepts.py#L44) won't be executed. I will download your filtered folder and have another run. Thank you!
